### ISSUE TITLE
Add support for Computed Object Properties

### ISSUE THREAD

**Issue by alexsasharegan** [beautify-web/js-beautify#1200]
# Description
When formatting an object literal with a computed property, an undesired new line is introduced after the object's initial curly brace.

This issue was originally filed in the vscode extension's repo at this url: https://github.com/HookyQR/VSCodeBeautify/issues/143

# Input
The code looked like this before beautification:
```js
export default {
    filesById({ files }, getters) {
        return files.reduce((memo, f) => Object.assign(memo, { [f.id]: f }), {})
    },
}
```

# Expected Output
The code should not have changed after beautification:
```js
export default {
    filesById({ files }, getters) {
        return files.reduce((memo, f) => Object.assign(memo, { [f.id]: f }), {})
    },
}
```

# Actual Output
The  code actually looked like this after beautification:
```js
export default {
    filesById({ files }, getters) {
        return files.reduce((memo, f) => Object.assign(memo, {
            [f.id]: f }), {})
    },
}
```

# Steps to Reproduce

## Environment
- Operating System: OS X 10.11.6
- VS Code Version: 1.12.2
- beautify Version: 1.0.2 (vscode extension)

## Settings
Example:
```json
{
    "end_with_newline": true,
    "indent_char": " ",
    "indent_size": 4,
    "indent_with_tabs": false,
    "max_preserve_newlines": 2,
    "wrap_line_length": 120,
    "indent_body_inner_html": false,
    "indent_scripts": "normal",
    "wrap_attributes": "force-aligned",
    "newline_between_rules": false,
    "break_chained_methods": false,
    "comma_first": false,
    "keep_array_indentation": false,
    "keep_function_indentation": false,
    "operator_position": "after-newline",
    "space_before_conditional": true,
    "space_in_paren": false,
    "space_in_empty_paren": false,
    "js": {
        "brace_style": "collapse,preserve-inline"
    }
}
```

**Comment by bitwiseman** [comment]
@alexsasharegan 
Thanks for the clear report and clean inputs and outputs.

This is es6, right?   Could you provide a link to the pertinent part of the es6 spec?

**Comment by alexsasharegan** [comment]
I've never dug into the specs, but here's a start from MDN:
- [computed property names](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer#Computed_property_names)
- this mixes with [shorthand method definitions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer#Method_definitions)

I'll search around for a link to the spec. Is there one canonical ECMA site I should be looking at?

**Comment by alexsasharegan** [comment]
Yeah, this stuff is a little over my head. Perhaps here: https://www.ecma-international.org/ecma-262/6.0/#sec-method-definitions-static-semantics-hascomputedpropertykey

**Comment by bitwiseman** [comment]
@alexsasharegan - thanks for taking the time, that's plenty.
It is very helpful when someone goes to fix an issue like this to have links, they don't have to be precise - close it good enough.  😄

**Comment by alexsasharegan** [comment]
Well, I've been curious to read up on the ECMA proposals and such, but the spec is basically legalese to me! Happy hunting on this one!

**Comment by alexsasharegan** [comment]
@bitwiseman, here's another place you can run into computed object props: [computed object property names and destructuring (MDN)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Computed_object_property_names_and_destructuring)

**Comment by alexsasharegan** [comment]
Oh, it looks like you've added support for this already! Next release, or already available?

**Comment by bitwiseman** [comment]
I closed the other one a duplicate.  This one should stay open.

**Comment by alexsasharegan** [comment]
It's been a year, I'm closing any issues of mine that are this inactive. I'll just remove my subscription then if you're still referencing this.

**Comment by bitwiseman** [comment]
This feature is block on javascript parsing.  The beautifier tokenizes but doesn't parse the input before beautifying.  Detecting computed properties properly requires significant look-ahead that is only reasonably accomplished by parsing before beautifying.  Even partial parsing might make this feasible.

**Comment by bitwiseman** [comment]
When implemented this feature should support correct `brace_style` behavior for destructured objects.

**Comment by sarora2073** [comment]
+1 on this. 

re: why it's blocked, perhaps it makes sense to have another brace-style option all together that is more blunt / simple in approach and just preserves the brace style period.
